nn.EmbeddingBag bound check

xw285cornell opened the pull request:
Summary: Today if we're accessing out of bound embedding rows, it'll either go through or throw IMA. This is not ideal - adding bound checks. This will probably slow things down - need to benchmark it.

Test Plan:
TODO: add some tests

Tried a simple example and it's showing this:
```
aten/src/ATen/native/cuda/EmbeddingBag.cu:143: EmbeddingBag_updateOutputKernel_sum_mean: block: [0,0,0], thread: [0,1,0] Assertion `input[emb] < numRows` failed.
```

Differential Revision: D43810777

ngimel reviewed (approved):
Last time you didn't want this check because of perf impact :-)

xw285cornell commented:
> Last time you didn't want this check because of perf impact :-)

haha yeah :) after seeing so many IMAs I certainly changed my mind ;)

xw285cornell committed:
nn.EmbeddingBag bound check (#96022)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/96022

Today if we're accessing out of bound embedding rows, it'll either go through or throw IMA. This is not ideal - adding bound checks. This will probably slow things down - need to benchmark it.

Test Plan:
Tried a simple example and it's showing this:
```
aten/src/ATen/native/cuda/EmbeddingBag.cu:143: EmbeddingBag_updateOutputKernel_sum_mean: block: [0,0,0], thread: [0,1,0] Assertion `input[emb] < numRows` failed.
```

Benchmark results on A100:
* After:
```
bs=2048, hashsize=10000000, dim=256, pooling_factor=200, bw=421.22GB/s, duration=9.96s
bs=1024, hashsize=10000000, dim=128, pooling_factor=20, bw=511.10GB/s, duration=0.21s
bs=512, hashsize=10000000, dim=64, pooling_factor=5, bw=32.69GB/s, duration=0.20s
```
* before:
```
bs=2048, hashsize=10000000, dim=256, pooling_factor=200, bw=1077.71GB/s, duration=3.89s
bs=1024, hashsize=10000000, dim=128, pooling_factor=20, bw=540.39GB/s, duration=0.19s
bs=512, hashsize=10000000, dim=64, pooling_factor=5, bw=33.86GB/s, duration=0.19s
```

So it seems to have quite significant hit for the extra-large size esp when the pooling factor is very big. It's probably a rare case though - and the check prevents a common issue which probably worth the extra overhead.

Test:
It seems CUDA_KERNEL_ASSERT is not catchable due to the nature of CUDA kernel asserts. However we have CUDA_KERNEL_ASSERT2 + DSA CUDA kernel launch (DSA=device side assert) which provides catchable device assert. But this requires change API - so I'll probably skip in this diff and maybe try next diff.

Reviewed By: cpuhrsch, ngimel

Differential Revision: D43810777

fbshipit-source-id: ba1763feaf6f2e7345e606ca4396ffd902ea0144
